Apply color palette on render.

NaderCHASER opened the pull request:
* CustomTemplateImageryProvider is based on UrlTemplateImageryProvider. Two main differences are it accepts a colorPalette array along with min/max values to pass to the tiler. The tiler can use the min/max to "bake in" a raw value into the PNG tile through the red channel.
* GLSL modifications are a nightmare. Duplicating code was the only way I could get this to reliably work due to an inability to supply a texture argument to sampleAndBlend. Need insight on how to best approach this.
* Is creating the new Texture in ImageryLayer appropriate?

This PR needs help and I know it. I've been posting on the forum but have yet to receive any real response from Ceisum graphics programmers.

![screen shot 2017-02-06 at 11 42 30 am](https://cloud.githubusercontent.com/assets/913068/22661396/7e2abdf8-ec62-11e6-956c-0eb2a3d02636.jpg)
![screen shot 2017-02-02 at 9 33 48 pm](https://cloud.githubusercontent.com/assets/913068/22661409/85ca9b00-ec62-11e6-8f0e-6a84fc029840.jpg)

Here's a task list:
- [x] WEBGL RENDER WARNING: there is no texture bound to the unit #
- [ ] Palette has to be assigned after load. Can this be moved to the options somehow?
- [x] ~~Palette weirdness on zooming (see attachments)~~ Determined to be a data issue. Fixed URL at c00a7b4
- [ ] When the Imagery Layer is removed, the edge of the Earth still has the last color palette applied. Does the texture need to be destroyed somehow?
- [ ] Determine if colorPalette texture is being created in proper location (`ImageryLayer.js`) and assigned to the proper variable (should it be private/separate from the option?)
- [ ] Using the Imagery Layer's `min`/`max`, add `pickFeatures` capabilities to display the value at a particular point. (Should it use `ImageryLayer._imageryCache`?)
- [ ] Write tests


https://groups.google.com/forum/#!msg/cesium-dev/CE_91W9dPUA/g9DrqgPJBAAJ
https://groups.google.com/forum/#!topic/cesium-dev/J3ws_ikULrk
https://groups.google.com/forum/#!searchin/cesium-dev/Custom$20image$20filters$20in$20SingleTileImageryProvider%7Csort:relevance/cesium-dev/aysBQWn9c88/JoR0dpQ_DAAJ

NASA JPL's use-case for this:
![at9_tld6](https://cloud.githubusercontent.com/assets/913068/22802176/16174390-eecd-11e6-8152-59f522733321.jpg)

hpinkos commented:
Thanks for the pull request @NaderCHASER!  Someone should be able to respond to this during our Bug Bash later this week.

pjcozzi commented:
Looks good, thanks @NaderCHASER!

Can you please send in a [CLA](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) so we can review this?

NaderCHASER commented:
@pjcozzi 

"You or any other entity (including a cross-claim or counterclaim in a 
      awsuit) alleging that your Contribution"

"lawsuit" is missing the "l". Filling it out and submitting it now.

NaderCHASER commented:
I have a server-side tiler up and running that can be used.

http://cache.allisonhouse.com/maps/raw.php?model=rtma&run=201702081600&hour=1486569600000&layers=tmp_2m_0&z=3&x=1&y=3&min=0&max=70

This is the current 2-meter temperature for the US. Min/max must be supplied for the tiler to determine the range to use in the red channel.

Here's a sample color palette: http://cache.allisonhouse.com/maps/palettes/TMP_2M.pal

Code for inside an XHR that retrieves the palette from above URL:
`            var lines = this.response.split('\n'),
                colors = {};

            Array.prototype.forEach.call(lines, function(line, i){
                var values = line.split(/\s+/);
                if(values.length >= 4) {
                    var key = parseFloat(values[0]);
                    colors[key] = [];
                    colors[key][0] = parseInt(values[1]);
                    colors[key][1] = parseInt(values[2]);
                    colors[key][2] = parseInt(values[3]);
                    colors[key][3] = (typeof(values[4]) !== 'undefined' && values[4] != '' ? parseInt(values[4]) : 255);
                    return true;
                }
            });

            palettes[name] = colors;`

Code that takes the palette data and creates a suitable array for use in a texture:
`                var colors = [];
                for (i in palettes[params.types.raster.name]) {
                    colors.push([i, palettes[params.types.raster.name][i]]);
                }
                colors.sort(function(a, b) {
                    return a[0] - b[0];
                });

                // ImageryProvider options
                options.palette = colors;
                options.min = colors[0][0];
                options.max = colors[colors.length-1][0];

                var colorTableData = new Uint8Array(4096);
                var palette = [];
                var max = null;
                var min = null;
                var diff = null;
                var currentVal = 0;
                var currentValI = 0;
                var prevVal = 0;
                var diffVal = 1.0;
                var nextColor = new Uint8Array(4);
                var prevColor = new Uint8Array(4);

                function setPalette(index, r, g, b, a) {
                    a = (typeof(a) !== 'undefined' ? a : 255);

                    var i = palette.length;

                    palette[i] = [];
                    palette[i]['takeoff'] = [];
                    palette[i]['approach'] = [];
                    palette[i]['value'] = parseFloat(index);

                    palette[i]['takeoff']['red'] = r;
                    palette[i]['takeoff']['green'] = g;
                    palette[i]['takeoff']['blue'] = b;
                    palette[i]['takeoff']['alpha'] = a;

                    if(i > 0) {
                        palette[i]['approach']['red'] = r;
                        palette[i]['approach']['green'] = g;
                        palette[i]['approach']['blue'] = b;
                        palette[i]['approach']['alpha'] = a;
                    }

                    if(index < min || min === null) {
                        min = parseFloat(index);
                    }

                    if(index > max || max === null) {
                        max = parseFloat(index);
                    }
                }

                for(var i = 0; i < colors.length; i++) {
                    setPalette(colors[i][0], colors[i][1][0], colors[i][1][1], colors[i][1][2], colors[i][1][3]);
                }

                diff = Math.abs(max - min);
                currentVal = min;
                prevVal = min;
                var stepSize = diff / 1024;

                for(var i = 16; i < 4096; i += 4) {
                    while (currentVal >= palette[currentValI]['value'] && (currentValI+1) < palette.length) {
                      prevColor[0] = palette[currentValI]['takeoff']['red'];
                      prevColor[1] = palette[currentValI]['takeoff']['green'];
                      prevColor[2] = palette[currentValI]['takeoff']['blue'];
                      prevColor[3] = palette[currentValI]['takeoff']['alpha'];
                      prevVal = palette[currentValI]['value'];
                      currentValI++;
                      diffVal = palette[currentValI]['value'] - prevVal;
                      nextColor[0] = palette[currentValI]['approach']['red'];
                      nextColor[1] = palette[currentValI]['approach']['green'];
                      nextColor[2] = palette[currentValI]['approach']['blue'];
                      nextColor[3] = palette[currentValI]['approach']['alpha'];
                    }
                    if (currentVal >= palette[currentValI]['value'] && (currentValI+1) >= palette.length) {
                        prevColor[0] = palette[currentValI]['approach']['red'];
                        prevColor[1] = palette[currentValI]['approach']['green'];
                        prevColor[2] = palette[currentValI]['approach']['blue'];
                        prevColor[3] = palette[currentValI]['approach']['alpha'];
                        prevVal = max;
                        currentVal = max;
                        diffVal = 1.0;
                        nextColor[0] = palette[currentValI]['approach']['red'];
                        nextColor[1] = palette[currentValI]['approach']['green'];
                        nextColor[2] = palette[currentValI]['approach']['blue'];
                        nextColor[3] = palette[currentValI]['approach']['alpha'];
                    }
                    colorTableData[i] = parseFloat(currentVal-prevVal) * (parseFloat(nextColor[0] - prevColor[0]) / diffVal) + prevColor[0];
                    colorTableData[i+1] = parseFloat(currentVal-prevVal) * parseFloat(nextColor[1] - prevColor[1]) / diffVal + prevColor[1];
                    colorTableData[i+2] = parseFloat(currentVal-prevVal) * parseFloat(nextColor[2] - prevColor[2]) / diffVal + prevColor[2];
                    colorTableData[i+3] = parseFloat(currentVal-prevVal) * parseFloat(nextColor[3] - prevColor[3]) / diffVal + prevColor[3];
                    currentVal += stepSize;
                }
            }

            options.colorPalette = colorTableData;
`

And here is what the rendered image should look like (but doesn't, see post #1):
![screen shot 2017-02-08 at 10 18 45 am](https://cloud.githubusercontent.com/assets/913068/22748676/0c26201c-ede8-11e6-98fb-ad0980884fd8.jpg)

bagnell commented:
I haven't been able to run an example. Can you post a Sandcastle gist? @kring would know where this fits better.

I read through the forum posts that you linked. Was the only problem in the second post that `sampleAndBlend` would not compile without passing the last `sampler2D`? Have you tried using `#ifdef`s to remove the parameter? Try

```javascript
vec4 sampleAndBlend(
    vec4 previousColor,
    sampler2D texture,
#ifdef APPLY_COLOR_PALETTE
    sampler2D textureColorPalette,
#endif
    vec2 tileTextureCoordinates,
    vec4 textureCoordinateRectangle,
    vec4 textureCoordinateTranslationAndScale,
    float textureAlpha,
    float textureBrightness,
    float textureContrast,
    float textureHue,
    float textureSaturation,
    float textureOneOverGamma,
    float split)
{
//...
```
and rearranging the call in GlobeSurfaceShaderSet.js.

NaderCHASER commented:
@bagnell I haven't tried the #ifdef in the parameter. I actually didn't know that was possible. This whole compiled GLSL (and just GLSL in general) is rather new to me.

I'm working on a sandcastle and the suggested change now. Will update in an hour or so once I've completed them.

NaderCHASER commented:
`APPLY_COLOR_PALETTE` will be set if there are **any** Imagery Layers which support a color palette. This means the argument will be bound in the shader and it will fail to compile due to not being able to provide a texture for certain layers.

I've committed a fix for an accidental code removal (frustumSplits[] in FrameState.js). I've also comitted a sandcastle. The problem is that I can't test the sandcastle myself due to a CORS issue. I'll be doing some server-side stuff to get a public URL which has "Access-Control-Allow-Origin: *"

NaderCHASER commented:
I've fixed the URL in the sandcastle. It also now responds to CORS appropriately.

Issues of note:
- [ ] WEBGL RENDER WARNING: there is no texture bound to the unit #
- [ ] Palette has to be assigned after load. Can this be moved to the options somehow?

bagnell commented on Source/Shaders/GlobeFS.glsl:
```diff
@@ -38,6 +38,10 @@ uniform float u_dayTextureOneOverGamma[TEXTURE_UNITS];
 #endif
 
+#ifdef APPLY_COLOR_PALETTE
+uniform sampler2D u_dayTextureColorPalette[TEXTURE_UNITS];
```
This is causing the WebGL warnings. There aren't as many color palettes as there are `TEXTURE_UNITS`. Define another constant for the number of color palettes and use that as the size of the array.

bagnell commented on Source/Shaders/GlobeFS.glsl:
```diff
@@ -67,6 +71,7 @@ varying vec3 v_mieColor;
 vec4 sampleAndBlend(
     vec4 previousColor,
     sampler2D texture,
+    sampler2D textureColorPalette,
```
Since you don have the same number of color palettes, create a uniform of indices into `u_dayTextureColorPalette` for each texture.

bagnell replied:
Or you can pass the index directly [here](https://github.com/AnalyticalGraphicsInc/cesium/pull/4953/files#diff-772d885056ceb2379c95a5dfc07d8d5aR190)

bagnell commented on Source/Shaders/GlobeFS.glsl:
```diff
@@ -110,6 +116,25 @@ vec4 sampleAndBlend(
     }
 #endif
 
+    if(applyTextureColorPalette > 0.0) {
```
If the index is a valid value, say `>=0` execute this code.

bagnell commented on Source/Shaders/GlobeFS.glsl:
```diff
@@ -110,6 +116,25 @@ vec4 sampleAndBlend(
+          a1 = mix(tA, tB, f.y);
+        }
+
+        vec4 pixColor = texture2D(textureColorPalette, vec2(0.0, a1));
```
Index directly into `u_dayTextureColorPalette` here.

NaderCHASER replied:
Using u_dayTextureColorPalette[index] here results in 

`'[]' : Index expression must be constant`

NaderCHASER commented:
I've pushed up a change which is an adaptation of your recommendation. I don't believe I can pass the index as an argument in GLSL because the array is of an unknown length at compile time. Correct me if I'm wrong here, but I just couldn't ever get that to work.

I did, however, create a `PALETTE_UNITS` constant to correct the `bindTexture` errors. I still get very weird results. I can't tell if these are due to the GLSL, or something else.

NaderCHASER commented:
Code has been updated to pass all checks. Issue list is in the top post. 

Biggest current issues are a) removing the "paletted" layer causes the underlying imagery to be paletted at the top edge and b) zooming in/out can lead to a strip of tiles not being paletted.

pjcozzi commented:
Thanks for all the updates, @NaderCHASER!

@bagnell when you have the chance, perhaps after #4979 and the orthographic projection, can you do another review here?

cesium-concierge commented:
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/forum/#!msg/cesium-dev/CE_91W9dPUA/g9DrqgPJBAAJ
https://groups.google.com/forum/#!topic/cesium-dev/J3ws_ikULrk
https://groups.google.com/forum/#!searchin/cesium-dev/Custom$20image$20filters$20in$20SingleTileImageryProvider%7Csort

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/pull/4953 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.
---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:
